node/bindnode/node.go

```diff
@@ -0,0 +1,1112 @@
+		}
+		node := &_node{
+			schemaType: field.Type(),
+			val:        fval,
+		}
+		return node, nil
+	case *schema.TypeMap:
+		var kval reflect.Value
+		switch ktyp := typ.KeyType().(type) {
+		case *schema.TypeString:
+			kval = reflect.ValueOf(key)
+		default:
+			asm := &_assembler{
+				schemaType: ktyp,
+				val:        reflect.New(w.val.Type().Key()).Elem(),
+			}
+			if err := (*_assemblerRepr)(asm).AssignString(key); err != nil {
+				return nil, err
+			}
+			kval = asm.val
+		}
+		fval := w.val.MapIndex(kval)
+		if !fval.IsValid() { // not found
+			return nil, ipld.ErrNotExists{
+				// TODO
+			}
+		}
+		// TODO: Error/panic if fval.IsNil() && !typ.ValueIsNullable()?
+		// Otherwise we could have two non-equal Go values (nil map,
+		// non-nil-but-empty map) which represent the exact same IPLD
+		// node when the field is not nullable.
+		if typ.ValueIsNullable() {
+			if fval.IsNil() {
+				return ipld.Null, nil
+			}
+			fval = fval.Elem()
+		}
+		node := &_node{
+			schemaType: typ.ValueType(),
+			val:        fval,
+		}
+		return node, nil
+	case *schema.TypeUnion:
+		var idx int
+		var mtyp schema.Type
+		for i, member := range typ.Members() {
+			if member.Name().String() == key {
+				idx = i
+				mtyp = member
+				break
+			}
+		}
+		if mtyp == nil { // not found
+			return nil, ipld.ErrNotExists{
+				// TODO
+			}
+		}
+		haveIdx := int(w.val.FieldByName("Index").Int())
+		if haveIdx != idx { // mismatching type
+			return nil, ipld.ErrNotExists{
+				// TODO
+			}
+		}
+		mval := w.val.FieldByName("Value").Elem()
+		node := &_node{
+			schemaType: mtyp,
+			val:        mval,
+		}
+		return node, nil
+	}
+	return nil, ipld.ErrWrongKind{
+		TypeName:   w.schemaType.Name().String(),
+		MethodName: "LookupByString",
+		// TODO
+	}
+}
+
+func (w *_node) LookupByNode(key ipld.Node) (ipld.Node, error) {
+	panic("TODO: LookupByNode")
+}
+
+func (w *_node) LookupByIndex(idx int64) (ipld.Node, error) {
+	switch typ := w.schemaType.(type) {
+	case *schema.TypeList:
+		if idx < 0 || int(idx) >= w.val.Len() {
+			return nil, ipld.ErrNotExists{
+				// TODO
+			}
+		}
+		// TODO: bounds check
+		val := w.val.Index(int(idx))
+		if typ.ValueIsNullable() {
+			if val.IsNil() {
+				return ipld.Null, nil
+			}
+			val = val.Elem()
+		}
+		return &_node{schemaType: typ.ValueType(), val: val}, nil
+	}
+	return nil, ipld.ErrWrongKind{
+		TypeName:   w.schemaType.Name().String(),
+		MethodName: "LookupByIndex",
+		// TODO
+	}
+}
+
+func (w *_node) LookupBySegment(seg ipld.PathSegment) (ipld.Node, error) {
+	panic("TODO: LookupBySegment")
+}
+
+func (w *_node) MapIterator() ipld.MapIterator {
+	switch typ := w.schemaType.(type) {
+	case *schema.TypeStruct:
+		return &_structIterator{
+			schemaType: typ,
+			fields:     typ.Fields(),
+			val:        w.val,
+		}
+	case *schema.TypeUnion:
+		return &_unionIterator{
+			schemaType: typ,
+			members:    typ.Members(),
+			val:        w.val,
+		}
+	case *schema.TypeMap:
+		panic("TODO: ")
```

Comment:
this one probably does need to be in this PR

Reply:
It's hard to draw the line. This PR is huge as is, so I drew the line at "what is necessary to make the tests pass". I plan to continue fixing TODOs as I add more tests in future PRs. If we decide that this many TODOs are not acceptable in master, then I'd pretty easily slip into 5k-line PR territory :)